compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -277,7 +277,9 @@ fn encode_region<'tcx>(region: Region<'tcx>, dict: &mut FxHashMap<DictKey<'tcx>,
             s.push('E');
             compress(dict, DictKey::Region(region), &mut s);
         }
-        // FIXME(@lcnr): Why is `ReEarlyParam` reachable here.
+        // This is possible becasuse trait objects created by new_dynamic_trait and
```

Comment:
I don't think this is valid justification. We should definitely not be encountering *any* regions rather than `LateBound` and `Erased` during codegen.

Reply:
This is also related to the information I provided above. Since we can can only include at most as much information that would be available at the call sites, otherwise the labels won't match, the maximum amount of information we can include at the call site for calls on trait objects we can obtain is by resolving the trait identity of the trait that implements the method (i.e., the unsubstituted/unmonomorphized trait/the trait as it is declared), and these include `ReEarlyParam` and `Param`.

Reply:
Yeah but... why are you trying to strictly match lifetimes for the CFI encoding? I was not aware rustc guaranteed it would not e.g. unify two function defs with the same body but different lifetime bounds, thus guaranteeing those would be distinct things to call after compilation?

Reply:
Yes, if two functions with different lifetime bounds may be folded, I think we should probably remove/skip lifetime encoding. I can send it in another PR (or as an additional commit here).

Reply:
Probably best to land it separately, this is already a fairly big PR as it is.